### DISCUSSION TITLE
From exchange-graphcache@3.3.1 Invalid key warning in console

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#1199] [Q&A] [answered]
After upgrading from `exchange-graphcache@3.3.0` to `exchange-graphcache@3.3.1` in my console this warning comes out thousands of times (because I have many lists with many items):

```
Invalid key: The GraphQL query at the field at `Player:256.EmailAddresses` has a selection set, but no key could be generated for the data at this field.
You have to request `id` or `_id` fields for all selection sets or create a custom `keys` config for `EmailAddresses`.
Entities without keys will be embedded directly on the parent entity. If this is intentional, create a `keys` config for `EmailAddresses` that always returns null.
(Caused At: Unnamed query, "PLAYER_FRAGMENT" Fragment)
```

It's correct?

**Comment by frederikhors** [comment]
I'm asking this because I already have my `keys` object:

```js
const keys = {
  EmailAddresses: null
}
```

**Comment by kitten** [comment, marked as answer]
This used to unintentionally, work, i.e.

> create a `keys` config for `EmailAddresses` that always returns null.

means you'll want `EmailAddresses: () => null` and not `EmailAddresses: null` as the typings of this config suggest :)

**Reply by frederikhors** [reply]
It works. I ♥ you.

**Reply by AngelEvenset** [reply]
I'm having an error very much like that, can you please tell me where to place `EmailAddresses: () => null` ?

**Reply by kitten** [reply]
@AngelEvenset: the error basically tells you exactly what to do 😅 however, if the error isn't making much sense, please do read the docs page on Graphcache -- at least reading the first page gives you a lot of foundational knowledge and also contains a section on keys: https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#custom-keys-and-non-keyable-entities

**Reply by myieye** [reply]
@kitten It also took me about 20 min to figure out where I needed to make the change.
I initially thought I had to update my response payload. Maybe I'm just stupid. 🤷
The error says nothing about caching or where the config is that I need to update.

I think it would be a big improvement to change it to something like:
> If this is intentional, add `'EmailAddresses': () => null` to the `keys` of your `cacheExchange` config.

Also, the link you posted above was much more helpful than the link that was included in my error (https://bit.ly/2XbVrpR#15). I think the contents at https://bit.ly/2XbVrpR#15 should link to https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#custom-keys-and-non-keyable-entities.
